explore/marginal-gain/2.Marginal-Gain-Multiple-Mutations.py

```diff
@@ -0,0 +1,185 @@
+    '675': 'BRCA2',   # BRCA2, DNA repair associated
+    '238': 'ALK',     # anaplastic lymphoma receptor tyrosine kinase
+    '4221': 'MEN1',   # menin 1
+    '5979': 'RET',    # ret proto-oncogene
+}
+
+
+# ## Median absolute deviation feature selection
+
+# In[43]:
+
+def fs_mad(x, y):
+    """    
+    Get the median absolute deviation (MAD) for each column of x
+    """
+    scores = mad(x) 
+    return scores, np.array([np.NaN]*len(scores))
+
+
+# ## Define pipeline and Cross validation model fitting
+
+# In[44]:
+
+# Parameter Sweep for Hyperparameters
+param_grid = {
+    'classify__loss': ['log'],
+    'classify__penalty': ['elasticnet'],
+    'classify__alpha': [10 ** x for x in range(-3, 1)],
+    'classify__l1_ratio': [0],
+}
+
+pipeline = Pipeline(steps=[
+    ('imputer', Imputer()),    
+    ('select', SelectKBest(fs_mad)),
+    ('standardize', StandardScaler()),
+    ('classify', SGDClassifier(random_state=0, class_weight='balanced'))
+])
+
+cv_pipeline = {}
+for k in ['a','b']:
+    pg = param_grid.copy()
+    if k == 'a': pg['select__k'] = ['all']
+    elif k=='b': pg['select__k'] = [2000]
+    cv_pipeline[k] =  GridSearchCV(estimator=pipeline, param_grid=pg, scoring='roc_auc')
+
+
+# In[45]:
+
+def get_aurocs(X, y, pipeline, series):
+    """
+    Fit the classifier for the given mutation (y) and output predictions for it
+    """
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=0)
+    pipeline.fit(X=X_train, y=y_train)
+    y_pred_train = pipeline.decision_function(X_train)
+    y_pred_test = pipeline.decision_function(X_test)
+    cv_score_df = grid_scores_to_df(pipeline.grid_scores_)
+    series['mean_cv_auroc'] = cv_score_df.score.max()
+    series['training_auroc'] = roc_auc_score(y_train, y_pred_train)
+    series['testing_auroc'] = roc_auc_score(y_test, y_pred_test)
+    return series
+
+def grid_scores_to_df(grid_scores):
+    """
+    Convert a sklearn.grid_search.GridSearchCV.grid_scores_ attribute to 
+    a tidy pandas DataFrame where each row is a hyperparameter-fold combinatination.
+    """
+    rows = list()
+    for grid_score in grid_scores:
+        for fold, score in enumerate(grid_score.cv_validation_scores):
+            row = grid_score.parameters.copy()
+            row['fold'] = fold
+            row['score'] = score
+            rows.append(row)
+    df = pd.DataFrame(rows)
+    return df
+
+auroc_df = {}
+
+
+# In[46]:
+
+get_ipython().run_cell_magic('time', '', '# Train model a: covariates only.\nwarnings.filterwarnings("ignore") # ignore deprecation warning for grid_scores_\nrows = list()\nfor m in list(mutations):\n    series = pd.Series()\n    series[\'mutation\'] = m\n    series[\'symbol\'] = mutations[m]\n    rows.append(get_aurocs(X[\'a\'], Y[m], cv_pipeline[\'a\'], series))\nauroc_df[\'a\'] = pd.DataFrame(rows)\nauroc_df[\'a\'].sort_values([\'symbol\', \'testing_auroc\'], ascending=[True, False], inplace=True)')
```

Comment:
Do what you think is best. My opinion is that it's almost always better to be explicit and save the reader from having to remember a "labeling dictionary" of sorts.